refactor(apple): Upgrade to Swift 6.2 with concurrency checks

klochowicz opened the pull request:
This PR upgrades the Swift client from Swift 5 to Swift 6.2, addressing all
concurrency-related warnings and runtime crashes that come with Swift 6's
strict concurrency checking.

## Swift 6 Concurrency Primer

**`actor`** - A new reference type that provides thread-safe, serialised
access to mutable state. Unlike classes, actors ensure that only one piece of
code can access their mutable properties at a time. Access to actor
methods/properties requires await and automatically hops to the actor's
isolated executor.

**`@MainActor`** - An attribute that marks code to run on the main thread.
Essential for UI updates and anything that touches UIKit/AppKit. When a
class/function is marked @MainActor, all its methods and properties inherit
this isolation.

**`@Sendable`** - A protocol indicating that a type can be safely passed
across concurrency domains (between actors, tasks, etc.). Value types
(structs, enums) with Sendable stored properties are automatically Sendable.
Reference types (classes) need explicit @unchecked Sendable if they manage
thread-safety manually.

**`nonisolated`** - Opts out of the containing type's actor isolation. For
example, a nonisolated method in a @MainActor class can be called from any
thread without await. Useful for static methods or thread-safe operations.

**`@concurrent`** - Used on closure parameters in delegate methods. Indicates
the closure may be called from any thread, preventing the closure from
inheriting the surrounding context's actor isolation. Critical for callbacks
from system frameworks that call from background threads.

**Data Races** - Swift 6 enforces at compile-time (and optionally at runtime)
that mutable state cannot be accessed concurrently from multiple threads. This
eliminates entire classes of bugs that were previously only caught through
testing or production crashes.

## Swift Language Upgrade

- **Bump Swift 5 → 6.2**: Enabled strict concurrency checking throughout the
  codebase
- **Enable ExistentialAny (SE-0335)**: Adds compile-time safety by making
  protocol type erasure explicit (e.g., any Protocol instead of implicit
  Protocol)
- **Runtime safety configuration**: Added environment variables to log
  concurrency violations during development instead of crashing, allowing
  gradual migration

## Concurrency Fixes

### Actor Isolation

- **TelemetryState actor** (Telemetry.swift:10): Extracted mutable telemetry
  state into a dedicated actor to eliminate data races from concurrent access
- **SessionNotification @MainActor isolation** (SessionNotification.swift:25):
  Properly isolated the class to MainActor since it manages UI-related
  callbacks
- **IPCClient caching** (IPCClient.swift): Fixed actor re-entrance issues and
  resource hash-based optimisation by caching the client instance in Store

### Thread-Safe Callbacks

- **WebAuthSession @concurrent delegate** (WebAuthSession.swift:46): The
  authentication callback is invoked from a background thread by
  ASWebAuthenticationSession. Marked the wrapper function as @concurrent to
  prevent MainActor inference on the completion handler closure, then
  explicitly hopped back to MainActor for the session.start() call. This
  fixes EXC_BAD_INSTRUCTION crashes at _dispatch_assert_queue_fail.
- **SessionNotification @concurrent delegate**
  (SessionNotification.swift:131): Similarly marked the notification delegate
  method as @concurrent and used Task { @MainActor in } to safely invoke the
  MainActor-isolated signInHandler

### Sendable Conformances

- Added Sendable to Resource, Site, Token, Configuration, and other model
  types that are passed between actors and tasks
- **LogWriter immutability** (Log.swift): Made jsonData immutable to prevent
  capturing mutable variables in @Sendable closures

### Nonisolated Methods

- **Static notification display** (SessionNotification.swift:73): Marked
  showSignedOutNotificationiOS() as nonisolated since it's called from the
  Network Extension (different process) and only uses thread-safe APIs

Fixes #10674
Fixes #10675

klochowicz committed:
chore(apple): Bump Swift language version to 6.0

Update to the latest Swift language version with its strict concurrency checkes.

Address most issues that came up, in case of non-trivial work add
`@unchecked` and `nonisolated` workarounds for now.

klochowicz committed:
fix(apple): resolve Swift 6 runtime crash in WebAuthSession

Fixes a Swift 6 concurrency violation that caused runtime crashes during
OAuth authentication flow with ASWebAuthenticationSession.

Swift 6 enforces strict actor isolation at runtime, and unfortunately
not everything is properly reported at compile time, as shown below.

The crash occurred because:

1. WebAuthSession.signIn() was @MainActor isolated
2. Closures defined in @MainActor functions inherit MainActor isolation
3. ASWebAuthenticationSession calls completion handler from background thread
4. Crash: EXC_BAD_INSTRUCTION at _dispatch_assert_queue_fail (Swift 6
    runtime violation detected)

Closure actor isolation is determined by WHERE the closure is defined,
NOT what it captures. Extracting MainActor properties to Sendable locals
doesn't help if the closure is still defined in a MainActor context.

Mark closure as "@concurrent" and be explicit about running
session.start() back in the MainActor context.

klochowicz committed:
refactor(swift): Add DispatchQueue for Telemetry mutable state

klochowicz committed:
build(swift): Enable ExistentialAny feature flag (SE-0335) for explicit protocol types

While not part of Swift 6 by default, this upcoming feature provides
  additional compile-time safety by:
  1. Making type erasure explicit and visible
  2. Clarifying actor isolation boundaries in concurrent contexts
  3. Catching protocol type usage that could lead to subtle bugs
  4. Preventing performance issues from implicit existential types

Reference:
  https://github.com/swiftlang/swift-evolution/blob/main/proposals/0335-existential-any.md

klochowicz committed:
chore(apple): Suppress Swift 6 runtime concurrency assertion crashes

Configure environment variables to log concurrency violations instead of
crashing at runtime, while maintaining strict concurrency checking at
compile time. This allows gradual migration to Swift 6 concurrency model
without blocking development.

The commit adds SWIFT_IS_CURRENT_EXECUTOR_LEGACY_MODE_OVERRIDE and
SWIFT_UNEXPECTED_EXECUTOR_LOG_LEVEL to both the Xcode scheme and Info.plist files for the
 main app and network extensions.

klochowicz committed:
fix(apple): Resolve SessionNotification data race

Use @MainActor isolation and @concurrent delegate pattern to eliminate
unsafe concurrent access to mutable signInHandler closure.

klochowicz committed:
fix(apple): Resolve IPCClient actor re-entrance and cache invalidation issues

Cache IPCClient instance in Store to fix resource hash-based optimisation.
Remove Task wrapper from fetchResources() to prevent actor re-entrance.
Add Sendable conformances to Resource, Site, and related types for Swift 6.

klochowicz committed:
refactor(apple): Extract TelemetryState actor

klochowicz committed:
chore: Bump Swift to 6.2

klochowicz committed:
fix(apple): Resolve Swift 6 SendableClosureCaptures warning in LogWriter

Make jsonData immutable to prevent capturing mutable variables in
@Sendable closures, eliminating the concurrency warning when upgrading
to Swift 6.2.

thomaseizinger commented on swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift:
```diff
@@ -294,6 +296,20 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
     }
   }
 
+  func consumeStopReason(_ completionHandler: @Sendable (Data?) -> Void) {
```
I thought this way of passing errors got removed in #10603 ?

thomaseizinger reviewed (approved):
Nice work! Thank you for including the details in the commit messages. As a non-Swift engineer, it would be nice to have a quick overview of Swift concurrency in the PR description. Like what does `@Sendable` mean? What is a nonisolated property? etc

Also, bear in mind that the commit messages from the individual ones are not carried over to `main`.

jamilbk commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/IPCClient.swift:
What do you think of making IPCClient just an enum (so it can't be instantiated) and requiring the caller to pass a valid `session` into each function?

This would push the isolation requirements to the caller instead of creating a new isolation context.

Since we only ever have one IPCClient, I'm not sure we benefit from the ability to create instances of it.

thomaseizinger replied:
Swift has free-functions so if we don't have any fields on a class, we might as well get rid of the class/enum, right?

jamilbk commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Managers/VPNConfigurationManager.swift:
```diff
@@ -21,7 +21,9 @@ enum VPNConfigurationManagerError: Error {
+/// Only contains a 'let' property. NETunnelProviderManager handles its own synchronisation.
+public class VPNConfigurationManager: @unchecked Sendable {
   // Persists our tunnel settings
   let manager: NETunnelProviderManager
```
If we make `IPCClient` a stateless container (i.e. `enum`), then we could make `manager` a static var here.

Essentially, there's only ever one valid `NETunnelProviderManager` and `NETunnelProviderSession` instantiated - it might simplify the concurrency model just a tad to represent that in our abstractions.

What it seems we are doing is creating multiple instances of things that then try to access one shared thing (the `NETunnelProvider*` objects) and then doing backflips to manage access to the one shared thing. Perhaps we can reduce the abstractions around the shared thing and expose its thread safety-ness directly to callers?

jamilbk commented on swift/apple/FirezoneKit/Sources/FirezoneKit/Stores/Store.swift:
```diff
@@ -42,6 +42,9 @@ public final class Store: ObservableObject {
   private var vpnConfigurationManager: VPNConfigurationManager?
```
If we make the manager/session a singleton instance, then we wouldn't need a cached IPCClient in Store.

jamilbk reviewed (commented):
Nice! Found one little async fix and left some suggestions for further refactoring to simplify the concurrency model a bit.

thomaseizinger commented:
> > Nice work! Thank you for including the details in the commit messages. As a non-Swift engineer, it would be nice to have a quick overview of Swift concurrency in the PR description. Like what does `@Sendable` mean? What is a nonisolated property? etc
> > Also, bear in mind that the commit messages from the individual ones are not carried over to `main`.
> 
> sure, good idea! would you like me to add it somewhere to the docs? Or just make a long commit message in the PR description?

I think PR description is a good trade-off. It is purposely not in a living document so doesn't need to be updated, it captures our knowledge at the time the code was written and it is something that we can easily read during review.

I don't know about other people's dev setup but I am quite fond of the "Git Lens" feature Zed where I can see the commit message of a line just by hovering it. Makes it really easy to answer the question, "why is this attribute here?".

klochowicz committed:
chore: Remove unused stop reason

klochowicz committed:
fix(apple): Resolve iOS build failures for Swift 6.2 concurrency

Fix compilation error and warnings introduced by Swift 6.2's stricter
concurrency checking.

Changes:
- SessionNotification: Replace @concurrent with nonisolated on delegate
  method. Swift 6.2 restricts @concurrent to async nonisolated functions.
  The delegate method uses completion handler pattern and doesn't need
  main actor isolation since it properly wraps main-actor work in Task.

- DeviceMetadata: Add nonisolated(unsafe) to UIDevice accessor methods.
  Swift 6.2 marks UIDevice.current as @MainActor isolated, but UIDevice
  is thread-safe for property reads per Apple documentation. This keeps
  methods synchronous while satisfying Swift 6.2 concurrency rules.

All changes follow Swift 6.2 best practices and existing codebase
patterns for nonisolated(unsafe) usage.

klochowicz committed:
fix(apple): Replace deprecated String(cString:) with String(decoding:)

Swift 6.2 deprecates String(cString:) in favour of String(decoding:)
with explicit null termination handling.

Changes:
- BindResolvers: Find and truncate null terminator, convert CChar to
  UInt8, then decode as UTF-8. Swift 6.2 requires explicit handling
  of null termination and type conversion from Int8 to UInt8.

- Adapter: Fix sendable closure capture error by using class box
  pattern. Swift 6.2's stricter concurrency checking doesn't allow
  mutable variable capture in @Sendable closures. The ResolversBox
  provides reference semantics while preserving semaphore-based
  synchronisation.

Verified with iOS build: BUILD SUCCEEDED with no warnings.

klochowicz committed:
fix(apple): Handle MainActor isolation for UIDevice

Address Swift 6.2 concurrency warnings by properly handling MainActor-
isolated DeviceMetadata methods in Adapter.start().

Changes:
- Add LockedState utility for thread-safe state transfer (compatible
  with iOS 15+, unlike OSAllocatedUnfairLock which requires iOS 16+)
- Update Adapter.start() to call @MainActor methods via Task and use
  semaphore to wait synchronously for device metadata

This is a temporary solution. The proper fix is to refactor Adapter
to be an actor, which will naturally handle MainActor isolation and
eliminate the need for semaphore-based synchronisation.
